### PR TITLE
fix: Correctly support SelfType when searching for usages

### LINKED ISSUE #7443: Self in struct is not highlighted as usage
substrate-1611640509810

_Originally posted by @0x7CFE in https://github.com/rust-analyzer/rust-analyzer/issues/3887#issuecomment-767315005_

### PR DESCRIPTION
**Pull request by Veykril**
Fixes #7443

### CODE REVIEW

File: crates/ide/src/references.rs [hunk 1]
```diff
@@ -1166,18 +1166,67 @@ fn foo<const FOO$0: usize>() -> usize {
...
+        //     fn f() -> Self;
+        // }
+        // "#,
+        //             expect![[r#"
+        //                 Foo Trait FileId(0) 0..45 6..9
+
+        //                 FileId(0) 16..20
+        //                 FileId(0) 38..42
+        //             "#]],
+        //         );
```
Comment (Veykril) [inline crates/ide/src/references.rs:1192]: Are we somehow able to go from a `hir::Trait` to its corresponding SelfType?

Comment (matklad) [reply crates/ide/src/references.rs:1192]: We probably shouldn’t — in traits, Self is a type parameter. That is, if you want flag Self as usage of Foo, then you should flag T in `T: Foo` as well.

Comment (Veykril) [reply crates/ide/src/references.rs:1192]: Right, I didn't think of that

File: crates/ide_db/src/search.rs [hunk 2]
```diff
@@ -513,11 +531,31 @@ impl<'a> FindUsages<'a> {
...
+            ModuleDef::Adt(adt) => Some(adt.ty(sema.db)),
+            ModuleDef::TypeAlias(it) => Some(it.ty(sema.db)),
+            ModuleDef::BuiltinType(it) => {
+                let graph = sema.db.crate_graph();
+                let krate = graph.iter().next()?;
+                let root_file = graph[krate].root_file_id;
+                let module = sema.to_module_def(root_file)?;
+                Some(it.ty(sema.db, module))
+            }
+            ModuleDef::Trait(_it) => None, // FIXME turn trait into its self-type
```
[the remark this answers is not in the document]
Comment (Veykril) [reply crates/ide_db/src/search.rs]: Ye I didn't realize that, should definitely not do this for `Self` in trait defs then, changed the tests.

File: crates/ide/src/references.rs [hunk 3]
```diff
@@ -1166,18 +1166,67 @@ fn foo<const FOO$0: usize>() -> usize {
...
+}
+"#,
+            expect![[r#"
+                impl Impl FileId(0) 13..57 18..21
+
+                FileId(0) 18..21
+                FileId(0) 28..32
+                FileId(0) 50..54
+            "#]],
+        );
```
[the remark this answers is not in the document]
Comment (Veykril) [reply crates/ide/src/references.rs:1233]: The one downside of expect-testing is rustfmt not touch it while it also continues to pass with a different indentation 😄

**Pull request merged**